zkevm-circuits/src/evm_circuit/util/math_gadget.rs

```diff
@@ -816,3 +816,244 @@ impl<F: FieldExt, const NUM_BYTES: usize> MaxGadget<F, NUM_BYTES> {
         Ok(select::value(lt, rhs, lhs))
     }
 }
+#[derive(Clone, Debug)]
+pub(crate) struct DivWordsGadget<F> {
+    dividend: util::Word<F>,
+    divisor: util::Word<F>,
+    quotient: util::Word<F>,
+    remainder: [Cell<F>; 32],
+    v0:[Cell<F>; 9],
+    comparison_lo: ComparisonGadget<F, 16>,
+    comparison_hi: ComparisonGadget<F, 16>,
+}
+
+impl<F: FieldExt> DivWordsGadget<F> {
+    pub(crate) fn construct(
+        cb: &mut ConstraintBuilder<F>,
+        dividend: util::Word<F>,
+        divisor: util::Word<F>,
+    ) -> Self {
+        let quotient = cb.query_word();
+        let remainder = array_init::array_init(|_| cb.query_byte());
+        let v0 = array_init::array_init(|_| cb.query_byte());
+
+        let comparison_lo = ComparisonGadget::construct(
+            cb,
+            from_bytes::expr(&remainder[0..16]),
+            from_bytes::expr(&divisor.cells[0..16]),
+        );
+        let (lt_lo, eq_lo) = comparison_lo.expr();
+        let comparison_hi = ComparisonGadget::construct(
+            cb,
+            from_bytes::expr(&remainder[16..32]),
+            from_bytes::expr(&divisor.cells[16..32]),
+        );
+        let (lt_hi, eq_hi) = comparison_hi.expr();
+        let lt = select::expr(lt_hi, 1.expr(), eq_hi.clone() * lt_lo);
+        cb.require_equal(
+            "remainder < quotient",
+            1.expr(),
+            lt,
+        );
+
+        let mut dividend_limbs = vec![];
+        let mut divisor_limbs = vec![];
+        let mut quotient_limbs = vec![];
+        let mut remainder_limbs = vec![];
+        for idx in 0..4 {
+            let now_idx = (idx * 8) as usize;
+            dividend_limbs.push(from_bytes::expr(&dividend.cells[now_idx..now_idx + 8]));
+            divisor_limbs.push(from_bytes::expr(&divisor.cells[now_idx..now_idx + 8]));
+            quotient_limbs.push(from_bytes::expr(&quotient.cells[now_idx..now_idx + 8]));
+            remainder_limbs.push(from_bytes::expr(&remainder[now_idx..now_idx + 8]));
+        }
+
+        let t0 = divisor_limbs[0].clone() * quotient_limbs[0].clone();
```

Comment:
In fact, I think you can simplify this since `divisor * quotient + residue` is still within 256-bit range.